### PR TITLE
Circleci performance improvement

### PR DESCRIPTION
**Pull request by dippindots**
This improves the performance by doing these:
- Use `large` resource
- Enable parallelism (3 parallel instances)
- Separate the all-in-one validation command into individual test commands (one for each study)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -9,6 +9,8 @@ jobs:
   build:
     docker:
       - image: circleci/python:3.6
+    resource_class: large
+    parallelism: 3
```
Comment (inodb) [inline .circleci/config.yml:13]: @dippindots should we also apply this to the `validate_all_studies` job?

Comment (dippindots) [reply .circleci/config.yml:13]: sure, let me update this

File: .circleci/validate_changed_studies.sh [hunk 2]
```diff
@@ -42,17 +42,41 @@ if [[ $num_studies > 0 ]]; then
...
-  validation_command="$HOME/cbioportal/core/src/main/scripts/importer/./validateStudies.py -d $HOME/repo/ -l $list_csv -p $HOME/repo/.circleci/portalinfo -html $test_reports_location"
+  validation_command=""
+  for study in ${list_csv//,/ }
+  do
+      # append the first study
+      if [ "$validation_command" = "" ] ; then
+        validation_command="$HOME/cbioportal/core/src/main/scripts/importer/./validateStudies.py -d $HOME/repo/ -l $study -p $HOME/repo/.circleci/portalinfo -html $test_reports_location/$study"
+      else
+        # run each validation individually in the background
+        validation_command="${validation_command} & $HOME/cbioportal/core/src/main/scripts/importer/./validateStudies.py -d $HOME/repo/ -l $study -p $HOME/repo/.circleci/portalinfo -html $test_reports_location/$study"
```
[the remark this answers is not in the document]
Comment (inodb) [reply .circleci/validate_changed_studies.sh:53]: ah nvm i see you handled this below my bad

File: .circleci/validate_changed_studies.sh [hunk 3]
```diff
@@ -42,17 +42,41 @@ if [[ $num_studies > 0 ]]; then
...
-  else
-    echo "Errors found"
-    # move errors to ERRORS/ folder:
-    erred_studies=`grep -rnlz $test_reports_location -e 'Validation status.*Failed' `
+  eval "$validation_command"
+  # Waiting for all background processes finishes
+  while true; do
+    wait -n || {
+      code="$?"
+      echo "waiting for all processes finishes ...................."
```
Comment (inodb) [inline .circleci/validate_changed_studies.sh:62]:
```suggestion
      echo "waiting for all processes to finish ...................."
```

File: .circleci/validate_changed_studies.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (dippindots) [reply .circleci/validate_changed_studies.sh:53]: @inodb We do collect every failing studies information at the end: in the example we have this
```
====List of error studies:====

/home/circleci/test-reports/public/lung_nci_2022/lung_nci_2022-validation.html /home/circleci/test-reports/public/ov_tcga_pan_can_atlas_2018/ov_tcga_pan_can_atlas_2018-validation.html /home/circleci/test-reports/public/paad_tcga_pan_can_atlas_2018/paad_tcga_pan_can_atlas_2018-validation.html
```

**Review by inodb — APPROVED** [review]
LGTM! Thanks! I made some small comments, but not essential to fix in this PR

File: .circleci/validate_changed_studies.sh [hunk 4]
```diff
@@ -42,17 +42,41 @@ if [[ $num_studies > 0 ]]; then
...
+    }
+  done;
+
+  # find all studies with error
+  erred_studies=`grep -rnlz $test_reports_location -e 'Validation status.*Failed' `
+  if [[ $? -eq 0 ]]; then
+    echo $'\n====List of error studies:====\n'
+    echo $erred_studies
     mv $erred_studies $test_reports_location/ERRORS
     exit 1
```
[the remark this answers is not in the document]
Comment (dippindots) [reply .circleci/validate_changed_studies.sh:76]: I am not sure about this... Is that possible?

**Comment by rmadupuri** [comment]
@dippindots thanks so much for fixing this!

Small thing, the log shows multiple warning messages and exit status messages for the same study. Is that expected when running processes in parallel? 

Screen Shot 2022-10-27 at 2 36 12 PM

**Comment by dippindots** [comment]
@rmadupuri This is a little bit confusing, it's because we are running multiple validation tasks in the background at the same time, and each task would produce information messages like this. So messages from different studies mixed together at here, that's why we can see multiple validation messages together(I can't do anything about it because it's coming from the study validation script).
After we merge this pr, we should only looking at the summary messages at the bottom.

**Review by rmadupuri — APPROVED** [review]

File: .circleci/validate_changed_studies.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (inodb) [reply .circleci/validate_changed_studies.sh:76]: if u have some corner case that the python script doesn't handle, like an exception in the code there

**Comment by inodb** [comment]
@rmadupuri good point! @dippindots it should be possible to redirect the output if you add some statement like `> myjob_output.txt`. Then at the end you could either show the output altogether by using e.g. `cat` or uploading it as an artifact to the job, so people can manually inspect it themselves

**Pull request merged**